File: tif_customization/tif_customization/page/items_receiving_repo/items_receiving_repo.js
frappe.pages['items-receiving-repo'].on_page_load = function(wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'MR and PO',
		single_column: true
	});

	// Create page instance
	let page_instance = new ItemsReceivingRepoPage(page, wrapper);
	page_instance.init();
}

// Page class
function ItemsReceivingRepoPage(page, wrapper) {
	this.page = page;
	this.wrapper = wrapper;
	this.filters = {};
	this.data = [];
}

ItemsReceivingRepoPage.prototype.init = function() {
	this.make_filters();
	this.make_datatable();
	this.refresh();
}

ItemsReceivingRepoPage.prototype.make_filters = function() {
	let me = this;
	
	me.filter_area = $('<div class="filter-area" style="padding: 15px; background: #f8f9fa; border-radius: 4px; margin-bottom: 20px;"></div>').appendTo(me.page.main);
	
	// Filter row 1
	let filter_row1 = $('<div class="row" style="margin-bottom: 10px;"></div>').appendTo(me.filter_area);
	
	// Material Request filter
	$('<div class="col-md-3" style="padding: 5px;"></div>').html(`
		<label>Material Request</label>
		<input type="text" class="form-control material-request-filter" placeholder="Filter by MR">
	`).appendTo(filter_row1);
	
	// Item Code filter
	$('<div class="col-md-3" style="padding: 5px;"></div>').html(`
		<label>Item Code</label>
		<input type="text" class="form-control item-code-filter" placeholder="Filter by Item Code">
	`).appendTo(filter_row1);
	
	// Acknowledgment Status filter
	$('<div class="col-md-3" style="padding: 5px;"></div>').html(`
		<label>Acknowledgment Status</label>
		<select class="form-control status-filter">
			<option value="">All</option>
			<option value="Pending">Pending</option>
			<option value="Acknowledged">Acknowledged</option>
		</select>
	`).appendTo(filter_row1);
	
	// Filter row 2
	let filter_row2 = $('<div class="row" style="margin-bottom: 10px;"></div>').appendTo(me.filter_area);
	
	// From Date
	$('<div class="col-md-3" style="padding: 5px;"></div>').html(`
		<label>From Date</label>
		<input type="date" class="form-control from-date-filter">
	`).appendTo(filter_row2);
	
	// To Date
	$('<div class="col-md-3" style="padding: 5px;"></div>').html(`
		<label>To Date</label>
		<input type="date" class="form-control to-date-filter">
	`).appendTo(filter_row2);
	
	// Action buttons
	$('<div class="col-md-6" style="padding: 5px; padding-top: 28px;"></div>').html(`
		<button class="btn btn-primary btn-sm apply-filters" style="margin-right: 10px;">
			<i class="fa fa-filter"></i> Apply Filters
		</button>
		<button class="btn btn-secondary btn-sm clear-filters">
			<i class="fa fa-times"></i> Clear
		</button>
	`).appendTo(filter_row2);
	
	// Summary cards
	let summary_row = $('<div class="row" style="margin-top: 15px;"></div>').appendTo(me.filter_area);
	
	// Total Documents card
	me.summary_total_docs = $('<div class="col-md-4" style="padding: 5px;"></div>').html(`
		<div class="card" style="background: #e3f2fd; padding: 15px; border-radius: 4px;">
			<h5 style="margin: 0; color: #1976d2;">Total Documents</h5>
			<h3 style="margin: 5px 0 0 0; color: #1976d2;" class="summary-total-documents">0</h3>
		</div>
	`).appendTo(summary_row);
	
	// MR Count card
	me.summary_mr_count = $('<div class="col-md-4" style="padding: 5px;"></div>').html(`
		<div class="card" style="background: #fff3e0; padding: 15px; border-radius: 4px;">
			<h5 style="margin: 0; color: #f57c00;">MR Count</h5>
			<h3 style="margin: 5px 0 0 0; color: #f57c00;" class="summary-mr-count">0</h3>
		</div>
	`).appendTo(summary_row);
	
	// PO Count card
	me.summary_po_count = $('<div class="col-md-4" style="padding: 5px;"></div>').html(`
		<div class="card" style="background: #e8f5e9; padding: 15px; border-radius: 4px;">
			<h5 style="margin: 0; color: #388e3c;">PO Count</h5>
			<h3 style="margin: 5px 0 0 0; color: #388e3c;" class="summary-po-count">0</h3>
		</div>
	`).appendTo(summary_row);
	
	// Bind filter events - use me.filter_area to scope events
	me.filter_area.on('click', '.apply-filters', function() {
		me.apply_filters();
	});
	
	me.filter_area.on('click', '.clear-filters', function() {
		me.clear_filters();
	});
	
	// Enter key on filters
	me.filter_area.on('keypress', '.material-request-filter, .item-code-filter', function(e) {
		if (e.which === 13) {
			me.apply_filters();
		}
	});
}

ItemsReceivingRepoPage.prototype.apply_filters = function() {
	let me = this;
	
	me.filters = {
		material_request: $('.material-request-filter').val() || undefined,
		item_code: $('.item-code-filter').val() || undefined,
		acknowledgment_status: $('.status-filter').val() || undefined,
		from_date: $('.from-date-filter').val() || undefined,
		to_date: $('.to-date-filter').val() || undefined
	};
	
	// Remove undefined values
	Object.keys(me.filters).forEach(key => {
		if (me.filters[key] === undefined) {
			delete me.filters[key];
		}
	});
	
	me.refresh();
}

ItemsReceivingRepoPage.prototype.clear_filters = function() {
	let me = this;
	
	$('.material-request-filter').val('');
	$('.item-code-filter').val('');
	$('.status-filter').val('');
	$('.from-date-filter').val('');
	$('.to-date-filter').val('');
	
	me.filters = {};
	me.refresh();
}

ItemsReceivingRepoPage.prototype.make_datatable = function() {
	let me = this;
	
	me.datatable_area = $('<div class="datatable-area" style="overflow-x: auto; margin-top: 20px;"></div>').appendTo(me.page.main);
	
	// Create table structure
	me.table = $(`
		<table class="table table-bordered table-hover" style="width: 100%; min-width: 900px;">
			<thead style="background-color: #f8f9fa;">
				<tr>
					<th style="min-width: 120px;">Material Request</th>
					<th style="min-width: 100px;">MR Date</th>
					<th style="min-width: 120px;">Item Code</th>
					<th style="min-width: 150px;">Item Name</th>
					<th style="min-width: 110px;">Requested Qty</th>
					<th style="min-width: 110px;">Received Qty</th>
					<th style="min-width: 110px;">Pending Qty</th>
				</tr>
			</thead>
			<tbody class="table-body">
			</tbody>
		</table>
	`).appendTo(me.datatable_area);
	
	me.table_body = me.table.find('.table-body');
	
	// Bind acknowledge button click
	me.datatable_area.on('click', '.acknowledge-btn', function() {
		const material_request_item = $(this).data('item');
		me.acknowledge_item(material_request_item);
	});
}

ItemsReceivingRepoPage.prototype.render_table = function(data) {
	let me = this;
	
	me.table_body.empty();
	
	if (!data || data.length === 0) {
		me.table_body.append(`
			<tr>
				<td colspan="7" class="text-center" style="padding: 40px; color: #999;">
					${__('No MR Items found')}
				</td>
			</tr>
		`);
		return;
	}
	
	data.forEach(function(row) {
		const mr_link = `<a href="/app/material-request/${row.material_request}" target="_blank">${row.material_request}</a>`;
		const mr_date = row.mr_date ? frappe.datetime.str_to_user(row.mr_date) : '-';
		const requested_qty = `${flt(row.requested_qty).toFixed(2)} ${row.uom || ''}`;
		const received_qty = `${flt(row.received_qty).toFixed(2)} ${row.uom || ''}`;
		const pending_qty = flt(row.pending_qty);
		const pending_qty_html = `<span style="color: ${pending_qty > 0 ? '#f57c00' : '#388e3c'}; font-weight: bold;">${pending_qty.toFixed(2)} ${row.uom || ''}</span>`;
		
		const tr = $(`
			<tr>
				<td>${mr_link}</td>
				<td>${mr_date}</td>
				<td>${row.item_code || ''}</td>
				<td>${row.item_name || ''}</td>
				<td>${requested_qty}</td>
				<td>${received_qty}</td>
				<td>${pending_qty_html}</td>
			</tr>
		`);
		
		me.table_body.append(tr);
	});
}

ItemsReceivingRepoPage.prototype.refresh = function() {
	let me = this;
	
	// Show loading in table
		me.table_body.html(`
			<tr>
				<td colspan="7" class="text-center" style="padding: 40px; color: #999;">
					<i class="fa fa-spinner fa-spin"></i> ${__('Loading...')}
				</td>
			</tr>
		`);
	
	frappe.call({
		method: 'tif_customization.tif_customization.page.items_receiving_repo.items_receiving_repo.get_mr_items_receiving_data',
		args: {
			filters: me.filters
		},
		callback: function(r) {
			if (r.message && r.message.error) {
				me.table_body.html(`
					<tr>
						<td colspan="7" class="text-center text-danger" style="padding: 40px;">
							${__('Error loading data: {0}', [r.message.error])}
						</td>
					</tr>
				`);
				frappe.show_alert({
					message: __('Error loading data: {0}', [r.message.error]),
					indicator: 'red'
				}, 5);
				return;
			}
			
			if (r.message && r.message.data) {
				me.data = r.message.data;
				me.render_table(me.data);
				
				// Update summary cards
				$('.summary-total-documents').text(r.message.total_documents || 0);
				$('.summary-mr-count').text(r.message.mr_count || 0);
				$('.summary-po-count').text(r.message.po_count || 0);
			}
		}
	});
}

ItemsReceivingRepoPage.prototype.acknowledge_item = function(material_request_item) {
	let me = this;
	
	// Show dialog for remarks
	let d = new frappe.ui.Dialog({
		title: __('Acknowledge Receiving'),
		fields: [
			{
				fieldtype: 'Small Text',
				fieldname: 'remarks',
				label: __('Remarks (Optional)'),
				default: ''
			}
		],
		primary_action_label: __('Acknowledge'),
		primary_action(values) {
			d.hide();
			me.submit_acknowledgment(material_request_item, values.remarks || '');
		}
	});
	
	d.show();
}

ItemsReceivingRepoPage.prototype.submit_acknowledgment = function(material_request_item, remarks) {
	let me = this;
	
	// Show loading alert
	frappe.show_alert({
		message: __('Acknowledging...'),
		indicator: 'blue'
	}, 2);
	
	frappe.call({
		method: 'tif_customization.tif_customization.page.items_receiving_repo.items_receiving_repo.acknowledge_mr_item',
		args: {
			material_request_item: material_request_item,
			remarks: remarks
		},
		callback: function(r) {
			if (r.message && r.message.status === 'success') {
				frappe.show_alert({
					message: __('Item acknowledged successfully'),
					indicator: 'green'
				}, 3);
				
				// Refresh data immediately
				me.refresh();
			} else {
				frappe.show_alert({
					message: __('Error: {0}', [r.message.message || 'Unknown error']),
					indicator: 'red'
				}, 5);
			}
		}
	});
}

// Helper function for number formatting
function flt(value) {
	return parseFloat(value) || 0;
}

File: tif_customization/tif_customization/page/items_receiving_repo/items_receiving_repo.py
import frappe
from frappe import _
from frappe.utils import flt, getdate


@frappe.whitelist()
def get_mr_items_receiving_data(filters=None):
	"""Get MR Items with acknowledgment status"""
	try:
		if isinstance(filters, str):
			import json
			filters = json.loads(filters)
		elif not filters:
			filters = {}
		
		# Build filters
		conditions = ["mr.docstatus = 1"]  # Only submitted MRs
		params = {}
		
		# Material Request filter
		if filters.get('material_request'):
			conditions.append("mr.name = %(material_request)s")
			params['material_request'] = filters['material_request']
		
		# Item Code filter
		if filters.get('item_code'):
			conditions.append("mri.item_code = %(item_code)s")
			params['item_code'] = filters['item_code']
		
		# Date Range filter
		if filters.get('from_date'):
			conditions.append("mr.transaction_date >= %(from_date)s")
			params['from_date'] = filters['from_date']
		
		if filters.get('to_date'):
			conditions.append("mr.transaction_date <= %(to_date)s")
			params['to_date'] = filters['to_date']
		
		# Query to get MR Items with acknowledgment status
		# Use subquery to properly filter by acknowledgment status
		ack_status_condition = ""
		if filters.get('acknowledgment_status'):
			if filters['acknowledgment_status'] == 'Pending':
				ack_status_condition = "AND ack.name IS NULL"
			elif filters['acknowledgment_status'] == 'Acknowledged':
				ack_status_condition = "AND ack.name IS NOT NULL"
		
		query = f"""
			SELECT 
				mr.name AS material_request,
				mr.transaction_date AS mr_date,
				mri.name AS material_request_item,
				mri.item_code,
				mri.item_name,
				mri.qty AS requested_qty,
				mri.stock_qty AS requested_stock_qty,
				mri.uom,
				mri.stock_uom,
				COALESCE(SUM(pr_item.qty), 0) AS received_qty,
				COALESCE(SUM(pr_item.stock_qty), 0) AS received_stock_qty,
				poi.parent AS purchase_order,
				CASE 
					WHEN ack.status = 'Acknowledged' THEN 'Acknowledged'
					ELSE 'Pending'
				END AS acknowledgment_status,
				ack.acknowledged_date AS acknowledgment_date,
				ack.acknowledged_by AS acknowledged_by,
				ack.name AS acknowledgment_name
			FROM `tabMaterial Request` mr
			INNER JOIN `tabMaterial Request Item` mri ON mri.parent = mr.name
			INNER JOIN `tabAcknowledgment` ack ON ack.material_request = mr.name 
				AND ack.docstatus != 2
			LEFT JOIN `tabPurchase Order Item` poi ON poi.material_request_item = mri.name
			LEFT JOIN `tabPurchase Receipt Item` pr_item ON pr_item.purchase_order_item = poi.name 
				AND pr_item.docstatus = 1
			LEFT JOIN `tabPurchase Receipt` pr ON pr.name = pr_item.parent AND pr.docstatus = 1
			WHERE {' AND '.join(conditions)} {ack_status_condition}
			GROUP BY mr.name, mri.name, mri.item_code, mri.item_name, mri.qty, mri.stock_qty, 
				mri.uom, mri.stock_uom, mr.transaction_date, poi.parent, ack.status, 
				ack.acknowledged_date, ack.acknowledged_by, ack.name
			ORDER BY mr.transaction_date DESC, mr.name DESC, mri.item_code
		"""
		
		results = frappe.db.sql(query, params, as_dict=True)
		
		# Calculate pending qty and format data
		for row in results:
			# Use stock_qty for calculations if available, else use qty
			requested = flt(row.get('requested_stock_qty') or row.get('requested_qty'))
			received = flt(row.get('received_stock_qty') or row.get('received_qty'))
			row['pending_qty'] = requested - received
			row['pending_qty'] = max(0, row['pending_qty'])  # Don't show negative
			
			# Format dates
			if row.get('acknowledgment_date'):
				row['acknowledgment_date'] = str(row['acknowledgment_date'])
			if row.get('mr_date'):
				row['mr_date'] = str(row['mr_date'])
		
		# Get unique MR and PO counts
		unique_mrs = set()
		unique_pos = set()
		
		for row in results:
			if row.get('material_request'):
				unique_mrs.add(row['material_request'])
			if row.get('purchase_order'):
				unique_pos.add(row['purchase_order'])
		
		return {
			'data': results,
			'total_count': len(results),
			'pending_count': len([r for r in results if r['acknowledgment_status'] == 'Pending']),
			'acknowledged_count': len([r for r in results if r['acknowledgment_status'] == 'Acknowledged']),
			'mr_count': len(unique_mrs),
			'po_count': len(unique_pos),
			'total_documents': len(unique_mrs) + len(unique_pos)
		}
	
	except Exception as e:
		frappe.log_error(f"Error in get_mr_items_receiving_data: {str(e)}", "MR Items Receiving Error")
		return {'error': str(e), 'data': []}


@frappe.whitelist()
def acknowledge_mr_item(material_request_item, remarks=""):
	"""Acknowledge receiving of MR Item"""
	try:
		# Get MR Item details
		mr_item = frappe.get_doc("Material Request Item", material_request_item)
		mr_doc = frappe.get_doc("Material Request", mr_item.parent)
		
		# Check if acknowledgment already exists
		existing_ack = frappe.db.exists(
			"Acknowledgment",
			{
				"material_request": mr_doc.name,
				"status": "Acknowledged",
				"docstatus": 1
			}
		)
		
		if existing_ack:
			# Update existing acknowledgment
			ack_doc = frappe.get_doc("Acknowledgment", existing_ack)
		else:
			# Get Purchase Receipt for this MR
			pr_list = frappe.db.sql("""
				SELECT DISTINCT pr.name
				FROM `tabPurchase Receipt` pr
				INNER JOIN `tabPurchase Receipt Item` pr_item ON pr_item.parent = pr.name
				INNER JOIN `tabPurchase Order Item` poi ON poi.name = pr_item.purchase_order_item
				WHERE poi.material_request_item = %s
				AND pr.docstatus = 1
				ORDER BY pr.creation DESC
				LIMIT 1
			""", (material_request_item,), as_dict=True)
			
			pr_name = pr_list[0].name if pr_list else None
			
			# Get Purchase Order
			po_list = frappe.db.sql("""
				SELECT DISTINCT po.name
				FROM `tabPurchase Order` po
				INNER JOIN `tabPurchase Order Item` poi ON poi.parent = po.name
				WHERE poi.material_request_item = %s
				AND po.docstatus = 1
				ORDER BY po.creation DESC
				LIMIT 1
			""", (material_request_item,), as_dict=True)
			
			po_name = po_list[0].name if po_list else None
			
			# Create new acknowledgment
			ack_doc = frappe.get_doc({
				"doctype": "Acknowledgment",
				"purchase_receipt": pr_name,
				"purchase_order": po_name,
				"material_request": mr_doc.name,
				"requested_by": mr_doc.owner,
				"status": "Pending",
				"items": []
			})
			
			# Add items from Purchase Receipt if available
			if pr_name:
				pr_doc = frappe.get_doc("Purchase Receipt", pr_name)
				for pr_item in pr_doc.items:
					if pr_item.material_request_item == material_request_item:
						ack_doc.append("items", {
							"item_code": pr_item.item_code,
							"item_name": pr_item.item_name,
							"qty": pr_item.qty,
							"warehouse": pr_item.warehouse,
							"uom": pr_item.uom
						})
			
			ack_doc.insert(ignore_permissions=True)
		
		# Update acknowledgment
		ack_doc.status = "Acknowledged"
		ack_doc.acknowledged_by = frappe.session.user
		ack_doc.acknowledged_date = frappe.utils.now_datetime()
		ack_doc.acknowledgment_remarks = remarks
		ack_doc.save(ignore_permissions=True)
		ack_doc.submit()
		
		return {
			"status": "success",
			"message": _("MR Item acknowledged successfully"),
			"acknowledgment": ack_doc.name
		}
	
	except Exception as e:
		frappe.log_error(f"Error acknowledging MR Item {material_request_item}: {str(e)}", "MR Item Acknowledgment Error")
		return {
			"status": "error",
			"message": str(e)
		}
